on-the-fly cmoriser for ACCESS native data (#2430)
ESMValGroup/ESMValCore: doc/quickstart/find_data.rst

</../esmvalcore/config/extra_facets/ipslcm-mappings.yml>`. These multi-variable
files must also undergo some data selection.

.. _read_access-esm:

ACCESS-ESM
^^^^^^^^^^

ESMValTool can read native `ACCESS-ESM <https://research.csiro.au/access/about/esm1-5/>`__
model output.

.. warning::

  This is the first version of ACCESS-ESM CMORizer for ESMValCore Currently, 
  only two variables (`tas`, and `pr`) are fully supported.

To read ACCESS-ESM native data, users need to add the directory file structure to 
`config_developer.yml`.

.. code-block:: yaml

  ACCESS:
    cmor_strict: false
    input_dir:
      default:
        - '{institute}/{dataset_name}/{exp}/{modeling_realm}/netCDF'
    input_file:
      default: '{dataset}.{special_attr}-*.nc'
    output_file: '{dataset}.{special_attr}'
    cmor_type: 'CMIP6'

>     cmor_default_table_prefix: 'CMIP6_'

Reviewer: Please use a format for config-developer.yml so that most user's do not need to modify this. Note that you can also specify lists for the entries in `input_dir` and `input_file`, and also add more DRS in addition to `default`.

Example:

```yaml
ICON:
  cmor_strict: false
  input_dir:
    default:
      - '{exp}'
      - '{exp}/outdata'
  input_file:
    default: '{exp}_{var_type}*.nc'
    another_drs:
      - '{exp}_{mip}.nc'
      - '{short_name}_{exp}_{mip}.nc'
  output_file: '{project}_{dataset}_{exp}_{var_type}_{mip}_{short_name}'
  cmor_type: 'CMIP6'
  cmor_default_table_prefix: 'CMIP6_'
```

Instead of showing the config-developer.yml contents here, it would be nicer to just show the input file format just like here:

https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/find_data.html#emac

It would also be super helpful to document the extra facets (like `special_attr`) here.

.. hint::

  We only provide one default `input_dir` since this is how ACCESS-ESM native data was 
  stored on NCI. Users can modify this path to match their local file structure.

  Default `input_file` will automatically load all the native ACCESS-ESM data(1850-2014).
  If you only need to evaluate a shorter period. Users can add a more specific time range

>   (`{dataset}.{special_attr}-YYYYMM*.nc` e.g.) to avoid the cost of loading unnecessary data.

Reviewer: The desired time range should never be hardcoded in the config-developer.yml file but set in the recipe using the [`timerange`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/overview.html#time-ranges) facet.

Thus, example dataset entries could look like this:

.. code-block:: yaml

>   dataset:

Reviewer: This doesn't look resolved to me. `ACCESS-ESM1-5` is not the institute, but rather the name of the model, right? So this probably would be a better fit for `dataset` (you can then rename your fix file from `cmor/_fixes/access/access_esm.py` to `cmor/_fixes/access/access_esm1_5.py` and possible extend this with more fix files if you want to look at other versions of the model. Then you can get rid of `institute` altogether.

    - {project: ACCESS, institute: ACCESS-ESM1-5, mip: Amon, dataset:ACCESS_ESM, dataset_name: HI-CN-05, 
      exp: history, modeling_realm: atm, special_attr: pa, start_year: 1986, end_year: 1986}

`dataset` and `dataset_name` are not redundant, `dataset` is for ESMValCore to search for CMORizer,

> `dataset_name` is to search for data.

Reviewer: This is very confusing. Are `dataset` and `dataset_name` somehow related (does a specific `dataset` always have the same `dataset_name`)? If this is the case, would it be an option to rename the CMORizer so that you only have to use one of the keys?

purpose.
For some variables, extra facets are necessary; otherwise ESMValCore cannot
read them properly.
Supported keys for extra facets are:

==================== ====================================== =================================
Key                  Description                            Default value if not specified
==================== ====================================== =================================
``raw_name``         Variable name of the variable in the   CMOR variable name of the
                     raw input file                         corresponding variable
``modeling_realm``   Realm attribute include `atm`, `ice`   No default (needs to be 
                     and `oce`                              specified in extra facets or
                                                            recipe if default DRS is used)
```special_attr``    A special attribute in the filename    No default
                     `ACCESS-ESM` raw data, it's related to 
                     frquency of raw data
``sub_dataset``      Part of the ACCESS-ESM raw dataset     No default
                     root, need to specify if you want to

>                      use the cmoriser                                                                       

Reviewer 2: Is it part of the root ? I don't think so. careful to not create confusion with the rootpath in the config file.

</../esmvalcore/config/extra_facets/ipslcm-mappings.yml>`. These multi-variable
files must also undergo some data selection.

.. _read_access-esm:

ACCESS-ESM
^^^^^^^^^^

ESMValTool can read native `ACCESS-ESM <https://research.csiro.au/access/about/esm1-5/>`__
model output.

.. warning::

  This is the first version of ACCESS-ESM CMORizer for ESMValCore Currently, 
  only two variables (`tas`, and `pr`) are fully supported.

This is an example of the directory file structure in `config_developer.yml`.

.. code-block:: yaml

  ACCESS:
    cmor_strict: false
    input_dir:
      default:
        - '{institute}/{sub_dataset}/{exp}/{modeling_realm}/netCDF'
    input_file:
      default: '{sub_dataset}.{special_attr}-*.nc'

>     output_file: '{sub_dataset}.{special_attr}'

Reviewer 3: ```suggestion
    output_file: '{institute}_{sub_dataset}_{special_attr}_{short_name}'
```
Could edit in the `config-developer` file so that the output preprocessed files can have more info like other output files, including short_name and institute and separated by underscores.

  Supported variables: ``pr``, ``ps``, ``psl``, ``rlds``, ``tas``, ``ta``, ``va``,
  ``ua``, ``zg``, ``hus``, ``clt``, ``rsus``, ``rlus``.

This is an example of the directory file structure in `config_developer.yml`.

.. code-block:: yaml

  ACCESS:
    cmor_strict: false
    input_dir:
      default:
        - '{institute}/{sub_dataset}/{exp}/{modeling_realm}/netCDF'
    input_file:
      default: '{sub_dataset}.{special_attr}-*.nc'
    output_file: '{institute}_{sub_dataset}_{special_attr}_{short_name}'
    cmor_type: 'CMIP6'
    cmor_default_table_prefix: 'CMIP6_'

.. hint::

  We only provide one default `input_dir` since this is how ACCESS-ESM native data was 
  stored on NCI. Users can modify this path to match their local file structure.

Thus, example dataset entries could look like this:

.. code-block:: yaml

  dataset:
    - {project: ACCESS, institute: ACCESS-ESM1-5, mip: Amon, dataset:ACCESS_ESM, sub_dataset: HI-CN-05, 
      exp: history, modeling_realm: atm, special_attr: pa, start_year: 1986, end_year: 1986}

`dataset` and `sub_dataset` are not redundant, `dataset` is for ESMValCore to search for CMORizer,

> which is always `ACCESS_ESM`, `sub_dataset` is dataset under root `ACCESS-ESM1-5`.

Reviewer: This can be removed if you replace `institute` with `dataset` (then it should be clear from the directory structure what both keys mean).

`special_attr` is a special attribute in the filename `ACCESS-ESM` raw data, it's related to frquency
of raw data.

> `modeling_realm` is a realm attribute, it include `atm`, `ice` and `oce`.

Reviewer: Please move these two and an entry for `sub_dataset` to the table. These are also extra facets.

Similar to any other fix, the ACCESS-ESM fix allows the use of :ref:`extra
facets<extra_facets>`.

> By default, the file :download:`emac-mappings.yml

Reviewer: ```suggestion
By default, the file :download:`access-mappings.yml
```

</../esmvalcore/config/extra_facets/access-mappings.yml>` is used for that
purpose.
For some variables, extra facets are necessary; otherwise ESMValCore cannot
read them properly.
Supported keys for extra facets are:

> ==================== ====================================== =================================

Reviewer: Please also add `calculate` to the list of supported keys, this is used in the extra facets file.

Key                  Description                            Default value if not specified
==================== ====================================== =================================
``raw_name``         Variable name of the variable in the   CMOR variable name of the
                     raw input file                         corresponding variable

> ``modeling_realm``   Realm attribute include `atm`, `ice`   Default realm of this variable 
>                      and `oce`      

Reviewer: ```suggestion
``modeling_realm``   Realm attribute include `atm`, `ice`   No default (needs to be 
                     and `oce`                              specified in extra facets or
                                                            recipe if default DRS is used)
```

Technically, since you're using `NativeDatasetFix.get_cube`, `raw_units` is also allowed (see https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/find_data.html#emac)
